File: Javascript/2019 카카오 겨울 인턴/n03.js
function loop(bans, idx, filterd) {
    let ret = {};
    if (filterd.length === bans.length) {
        let tmp = filterd.slice();
        tmp = tmp.sort().join('');
        ret[tmp] = true;
    } else {
        for (let i = 0; i < bans[idx].length; i++) {
            if (!filterd.includes(bans[idx][i])) {
                filterd.push(bans[idx][i]);
                let res = loop(bans, idx+1, filterd);
                Object.keys(res).forEach(function(e) {
                    ret[e] = true;
                });
                filterd.pop();
            }
        }
    }
    return ret;
}

function solution(user_id, banned_id) {
    let answer = 0;
    const re = (a, b) =>
        new RegExp(`^${b}$`).test(a);
    for (let i = 0; i < banned_id.length; i++) {
        banned_id[i] = banned_id[i].replace(/\*/g, '.');
    }
    let bans_list = [];
    banned_id.forEach(b => {
        bans_list.push(user_id.filter(u => re(u, b)));
    });
    answer = loop(bans_list, 0, []);
    return Object.keys(answer).length;
}

console.log(solution( // result 2
    ["frodo", "fradi", "crodo", "abc123", "frodoc"], 
    ["fr*d*", "abc1**"]
));
console.log(solution( // result 2
    ["frodo", "fradi", "crodo", "abc123", "frodoc"],
    ["*rodo", "*rodo", "******"]
));
console.log(solution( // result 3
    ["frodo", "fradi", "crodo", "abc123", "frodoc"],
    ["fr*d*", "*rodo", "******", "******"]
));

File: Python/kakao/2019-winter-intern/n03.py
import re

def loop(bans_list, id_list, res):
    ret = list()
    for i, bans in enumerate(bans_list):
        for ban in bans:
            if id_list[int(ban)] == 0:
                if len(bans_list) == 1:
                    tmp = sorted(res+[ban])
                    ret.append("".join(tmp))
                else:
                    id_list[int(ban)] = 1
                    r = loop(bans_list[:i]+bans_list[i+1:], id_list, res+[ban])
                    ret += r
                    id_list[int(ban)] = 0
    return list(set(ret))

def solution(user_id, banned_id):
    if len(user_id) == len(banned_id):
        return 1
    for i in range(len(banned_id)):
        banned_id[i] = banned_id[i].replace("*", ".")
    id_list = [0]*len(user_id)
    bans_list = list()
    for ban in banned_id:
        regex = re.compile(f'^{ban}$')
        tmp_list = []
        for i, user in enumerate(user_id):
            if regex.match(user):
                tmp_list.append(str(i))
        bans_list.append(tmp_list)
    ret = loop(bans_list, id_list, list())
    return len(ret)

if __name__ == "__main__":
    # ans = solution(["frodo", "fradi", "crodo", "abc123", "frodoc"],
    #                 ["fr*d*", "*rodo", "******", "******"])
    # print(ans)
    ans = solution(["frodo11", "fradi1", "crodo1", "abc12", "frod1", "frod2", "frod3", "frod4"],
                    ["*******", "******", "******", "*****", "*****", "*****", "*****", "*****"])
    print(ans)